File: auth/src/main/java/de/mel/auth/file/StandardFile.java
package de.mel.auth.file;

import de.mel.auth.tools.N;

import java.io.*;

/**
 * this is the default {@link File} wrapper for everything that is not recent android
 */
public class StandardFile extends AbstractFile<StandardFile> {
    private File file;

    public File getFile() {
        return file;
    }

    public StandardFile() {

    }

    @Override
    public String getPath() {
        return file.getPath();
    }

    public StandardFile(StandardFile parent, String name) {
        file = new File(parent.file, name);
    }

    public StandardFile(StandardFile originalFile) {
        this.file = new File(originalFile.file.getAbsolutePath());
    }

    @Override
    public String toString() {
        return file.toString();
    }

    public StandardFile(File parent, String name) {
        this.file = new File(parent, name);
    }

    public StandardFile(File file) {
        this.file = file;
    }

    public StandardFile(String path) {
        this(new File(path));
    }


    @Override
    public void setPath(String path) {
        file = new File(path);
    }

    @Override
    public String getName() {
        return file.getName();
    }

    @Override
    public String getAbsolutePath() {
        return file.getAbsolutePath();
    }

    @Override
    public boolean exists() {
        return file.exists();
    }

    @Override
    public boolean isFile() {
        return file.isFile();
    }

//    @Override
//    public boolean move(FFile target) {
//        return file.renameTo(new File(target.getAbsolutePath()));
//    }

    @Override
    public boolean isDirectory() {
        return file.isDirectory();
    }

    @Override
    public Long length() {
        return file.length();
    }

    @Override
    public StandardFile[] listFiles() {
        return N.arr.cast(file.listFiles(File::isFile), N.converter(StandardFile.class, StandardFile::new));

    }

    @Override
    public StandardFile[] listDirectories() {
        return N.arr.cast(file.listFiles(File::isDirectory), N.converter(StandardFile.class, StandardFile::new));

    }


    @Override
    public boolean delete() {
        return file.delete();
    }

    @Override
    public StandardFile getParentFile() {
        return new StandardFile(new File(file.getAbsolutePath()).getParentFile());
    }

    @Override
    public boolean mkdirs() {
        return file.mkdirs();
    }

    @Override
    public FileInputStream inputStream() throws FileNotFoundException {
        return new FileInputStream(file);
    }

    @Override
    public FileOutputStream outputStream() throws FileNotFoundException {
        return new FileOutputStream(file);
    }

    @Override
    public String getSeparator() {
        return File.separator;
    }

    @Override
    public boolean hasSubContent(StandardFile subFile) {
        return subFile != null && subFile.getAbsolutePath().startsWith(file.getAbsolutePath());
    }

    @Override
    public String getCanonicalPath() throws IOException {
        return file.getCanonicalPath();
    }

    @Override
    public Long getFreeSpace() {
        return file.getFreeSpace();
    }

    @Override
    public Long getUsableSpace() {
        return file.getUsableSpace();
    }

    @Override
    public Long lastModified() {
        return file.lastModified();
    }

    @Override
    public boolean createNewFile() throws IOException {
        return file.createNewFile();
    }

    @Override
    public StandardFile[] listContent() {
        return N.arr.cast(file.listFiles(), N.converter(StandardFile.class, StandardFile::new));
    }
}

File: auth/src/main/kotlin/de/mel/auth/file/StandardFile.kt
package de.mel.auth.file

import de.mel.Lok
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException

/**
 * this is the default [File] wrapper for everything that is not recent android
 */
class StandardFile : AbstractFile<StandardFile> {
    lateinit var file: File
        private set

    constructor() {}

    override var path: String
        get() = file.path
        set(path) {
            file = File(path)
        }

    override fun getParentFile(): IFile? = StandardFile(file.parentFile)
    override fun hasSubContent(subFile: IFile): Boolean = subFile.absolutePath.startsWith(absolutePath)

    override fun canRead(): Boolean = file.canRead()

    constructor(parent: StandardFile, name: String) {
        file = File(parent.file, name)
    }

    constructor(originalFile: StandardFile) {
        file = File(originalFile.file.absolutePath)
    }

    override fun toString(): String {
        return file.toString()
    }

    constructor(parent: File, name: String) {
        file = File(parent, name)
    }

    constructor(file: File) {
        this.file = file
    }

    constructor(path: String) : this(File(path)) {}

    override val absolutePath: String
        get() = file.absolutePath

    override fun exists(): Boolean {
        return file.exists()
    }

    override val isFile: Boolean
        get() = file.isFile


    //    @Override
//    public boolean move(FFile target) {
//        return file.renameTo(new File(target.getAbsolutePath()));
//    }
    override val isDirectory: Boolean
        get() = file.isDirectory

    override fun length(): Long {
        return file.length()
    }

    override fun listFiles(): Array<StandardFile> = file.listFiles { f -> f.isFile }?.map { StandardFile(it) }?.toTypedArray()
            ?: emptyArray()

    override fun listDirectories(): Array<StandardFile> = file.listFiles { f -> f.isDirectory }?.map { StandardFile(it) }?.toTypedArray()
            ?: emptyArray()


    override fun delete(): Boolean {
        return file.delete()
    }

    override fun mkdirs(): Boolean {
        // todo debug
        if (file.name == "i0")
            Lok.debug()
        return file.mkdirs()
    }

    @Throws(FileNotFoundException::class)
    override fun inputStream(): FileInputStream? {
        return FileInputStream(file)
    }

    override fun getName(): String = file.name

    @Throws(FileNotFoundException::class)
    override fun writer(): AbstractFileWriter? {
        return StandardFileWriter(file)
    }

    override val separator: String
        get() = File.separator

    fun hasSubContent(subFile: StandardFile?): Boolean {
        return subFile != null && subFile.absolutePath.startsWith(file!!.absolutePath)
    }

    @get:Throws(IOException::class)
    override val canonicalPath: String
        get() = file!!.canonicalPath

    override val freeSpace: Long
        get() = file.freeSpace

    override val usableSpace: Long
        get() = file.usableSpace

    override fun lastModified(): Long? {
        return file.lastModified()
    }

    @Throws(IOException::class)
    override fun createNewFile(): Boolean {
        return file.createNewFile()
    }


    override fun listContent(): Array<StandardFile> {
        return file.listFiles()?.map { StandardFile(it) }?.toTypedArray() ?: emptyArray()
    }
}